src/base_alloc/base_alloc.c

```diff
@@ -0,0 +1,228 @@
+/*
+ * Copyright (C) 2024 Intel Corporation
+ *
+ * Under the Apache License v2.0 with LLVM Exceptions. See LICENSE.TXT.
+ * SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+*/
+
+#include <assert.h>
+
+#include "base_alloc.h"
+#include "base_alloc_internal.h"
+#include "utils_common.h"
+#include "utils_concurrency.h"
+
+// minimum size of a single pool of the base allocator
+#define MINIMUM_POOL_SIZE (ba_os_get_page_size())
+
+// minimum number of chunks used to calculate the size of pools
+#define MINIMUM_CHUNK_COUNT (128)
+
+// alignment of the base allocator
+#define MEMORY_ALIGNMENT (8)
+
+typedef struct umf_ba_chunk_t umf_ba_chunk_t;
+typedef struct umf_ba_next_pool_t umf_ba_next_pool_t;
+
+// memory chunk of size 'chunk_size'
+struct umf_ba_chunk_t {
+    umf_ba_chunk_t *next;
+    char user_data[];
+};
+
+// metadata is set and used only in the main (the first) pool
+struct umf_ba_main_pool_meta_t {
+    size_t pool_size; // size of each pool (argument of each ba_os_alloc() call)
+    size_t chunk_size;         // size of all memory chunks in this pool
+    os_mutex_t *free_lock;     // lock of free_list
+    umf_ba_chunk_t *free_list; // list of free chunks
+#ifndef NDEBUG
+    size_t n_pools;
+    size_t n_allocs;
+    size_t n_chunks;
+#endif /* NDEBUG */
+};
+
+struct umf_ba_pool_t {
```

Comment:
you could add a general comment above this struct (and `umf_ba_next_pool_t` as well). The distinction between "main" and the rest of the pools could be more clear.

Reply:
Done